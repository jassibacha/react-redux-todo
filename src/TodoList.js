import React, { useState } from 'react';
import TodoForm from './TodoForm';
import Todo from './Todo';
import { v4 as uuid } from 'uuid';

const TodoList = () => {
    const [todos, setTodos] = useState([]);

    const renderTodos = () => {
        return (
            <div className="mt-3">
                {todos.map((todo) => (
                    <Todo
                        task={todo.task}
                        id={todo.id}
                        key={todo.id}
                        deleteTodo={() => deleteTodo(todo.id)}
                    />
                ))}
            </div>
        );
    };

    const deleteTodo = (todoId) => {
        // Run setTodos again
        setTodos((todos) =>
            // Filter the todos array from state
            todos.filter((todo) => {
                // If the ID matches, remove it!
                return todo.id !== todoId;
            })
        );
    };

    /** Add new todo to list. */
    const addTodo = (todo) => {
        let newTodo = { ...todo, id: uuid() };
        setTodos((todos) => [...todos, newTodo]);
    };

    return (
        <div className="Todos">
            <TodoForm addTodo={addTodo} />
            {renderTodos()}
        </div>
    );
};

export default TodoList;
