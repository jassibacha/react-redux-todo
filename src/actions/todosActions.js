export const add = (todoData) => {
    return {
        type: 'ADD_TODO',
        payload: todoData,
    };
};

export const remove = (id) => {
    return {
        type: 'REMOVE_TODO',
        payload: id,
    };
};
