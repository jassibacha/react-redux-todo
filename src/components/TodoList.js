import React from 'react';
import TodoForm from './TodoForm';
import Todo from './Todo';
import { v4 as uuid } from 'uuid';
import { useSelector, useDispatch } from 'react-redux';
import { add, remove } from '../actions/todosActions';

const TodoList = () => {
    const todos = useSelector((state) => state.todos);

    const dispatch = useDispatch();

    const renderTodos = () => {
        return (
            <div className="mt-3">
                {todos.map((todo) => (
                    <Todo
                        task={todo.task}
                        id={todo.id}
                        key={todo.id}
                        deleteTodo={() => deleteTodo(todo.id)}
                    />
                ))}
            </div>
        );
    };

    const deleteTodo = (todoId) => {
        dispatch(remove(todoId));
    };

    /** Add new todo to list. */
    const addTodo = (todo) => {
        const newTodo = { ...todo, id: uuid() };
        dispatch(add(newTodo));
    };

    return (
        <div className="Todos">
            <TodoForm addTodo={addTodo} />
            {renderTodos()}
        </div>
    );
};

export default TodoList;
