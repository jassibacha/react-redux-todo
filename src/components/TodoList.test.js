import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import TodoList from './TodoList';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { add, remove } from '../actions/todosActions';

const mockStore = configureStore([]);

describe('TodoList', () => {
    let store;

    beforeEach(() => {
        store = mockStore({
            todos: [],
        });
    });

    it('renders without crashing', function () {
        render(
            <Provider store={store}>
                <TodoList />
            </Provider>
        );
    });

    it('matches snapshot', function () {
        const { asFragment } = render(
            <Provider store={store}>
                <TodoList />
            </Provider>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('can add a new todo', function () {
        render(
            <Provider store={store}>
                <TodoList />
            </Provider>
        );

        // check there's no Delete button
        expect(screen.queryByText('X')).not.toBeInTheDocument();

        const taskInput = screen.getByPlaceholderText('Enter a new task');
        const submitBtn = screen.getByText('Add');

        // fill out the form
        fireEvent.change(taskInput, { target: { value: 'A Task' } });
        fireEvent.click(submitBtn);

        // Assert that an 'ADD_TODO' action was dispatched
        const actions = store.getActions();
        expect(actions[0].type).toBe('ADD_TODO');
        expect(actions[0].payload).toHaveProperty('task', 'A Task');
    });

    it('can delete a todo', function () {
        render(
            <Provider store={store}>
                <TodoList />
            </Provider>
        );

        // Directly dispatch add action
        const todoToAdd = { id: '123', task: 'A Task' };
        store.dispatch(add(todoToAdd));

        // Assert that an 'ADD_TODO' action was dispatched
        let actions = store.getActions();
        expect(actions[0].type).toBe('ADD_TODO');
        expect(actions[0].payload).toEqual(todoToAdd);

        // Now test deleting the todo
        store.dispatch(remove(todoToAdd.id));

        // Refresh actions array after dispatching delete
        actions = store.getActions();

        // Assert that a 'REMOVE_TODO' action was dispatched
        expect(actions[1].type).toBe('REMOVE_TODO');
        expect(actions[1].payload).toBe(todoToAdd.id);
    });
});
