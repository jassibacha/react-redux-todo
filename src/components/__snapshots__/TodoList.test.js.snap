// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Todos"
  >
    <div
      class="card bg-light mb-3"
    >
      <div
        class="card-body"
      >
        <form
          class="TodoForm row"
        >
          <div
            class="col-10"
          >
            <input
              class="form-control"
              id="task"
              name="task"
              placeholder="Enter a new task"
              type="text"
              value=""
            />
          </div>
          <div
            class="col-2"
          >
            <button
              class="btn btn-primary w-100"
              type="submit"
            >
              Add
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      class="mt-3"
    />
  </div>
</DocumentFragment>
`;
